flight/Modules/System/systemmod.c

```diff
@@ -618,6 +693,218 @@ void vApplicationStackOverflowHook(uintptr_t pxTask, signed char * pcTaskName)
+ */
+void EventGetStats(EventStats* statsOut)
+{
+	PIOS_Recursive_Mutex_Lock(mutex, PIOS_MUTEX_TIMEOUT_MAX);
+	memcpy(statsOut, &stats, sizeof(EventStats));
+	PIOS_Recursive_Mutex_Unlock(mutex);
+}
+
+/**
+ * Clear the statistics counters
+ */
+void EventClearStats()
+{
+	PIOS_Recursive_Mutex_Lock(mutex, PIOS_MUTEX_TIMEOUT_MAX);
+	memset(&stats, 0, sizeof(EventStats));
+	PIOS_Recursive_Mutex_Unlock(mutex);
+}
+
+/**
+ * Dispatch an event at periodic intervals.
+ * \param[in] ev The event to be dispatched
+ * \param[in] cb The callback to be invoked
+ * \param[in] periodMs The period the event is generated
+ * \return Success (0), failure (-1)
+ */
+/* XXX TODO: would be nice to get context record logic in these */
+int32_t EventPeriodicCallbackCreate(UAVObjEvent* ev, UAVObjEventCallback cb, uint16_t periodMs)
+{
+	return eventPeriodicCreate(ev, cb, 0, periodMs);
+}
+
+/**
+ * Update the period of a periodic event.
+ * \param[in] ev The event to be dispatched
+ * \param[in] cb The callback to be invoked
+ * \param[in] periodMs The period the event is generated
+ * \return Success (0), failure (-1)
+ */
+int32_t EventPeriodicCallbackUpdate(UAVObjEvent* ev, UAVObjEventCallback cb, uint16_t periodMs)
+{
+	return eventPeriodicUpdate(ev, cb, 0, periodMs);
+}
+
+/**
+ * Dispatch an event at periodic intervals.
+ * \param[in] ev The event to be dispatched
+ * \param[in] queue The queue that the event will be pushed in
+ * \param[in] periodMs The period the event is generated
+ * \return Success (0), failure (-1)
+ */
+int32_t EventPeriodicQueueCreate(UAVObjEvent* ev, struct pios_queue *queue, uint16_t periodMs)
+{
+	return eventPeriodicCreate(ev, 0, queue, periodMs);
+}
+
+/**
+ * Update the period of a periodic event.
+ * \param[in] ev The event to be dispatched
+ * \param[in] queue The queue
+ * \param[in] periodMs The period the event is generated
+ * \return Success (0), failure (-1)
+ */
+int32_t EventPeriodicQueueUpdate(UAVObjEvent* ev, struct pios_queue *queue, uint16_t periodMs)
+{
+	return eventPeriodicUpdate(ev, 0, queue, periodMs);
+}
+
+/**
+ * Dispatch an event through a callback at periodic intervals.
+ * \param[in] ev The event to be dispatched
+ * \param[in] cb The callback to be invoked or zero if none
+ * \param[in] queue The queue or zero if none
+ * \param[in] periodMs The period the event is generated
+ * \return Success (0), failure (-1)
+ */
+static int32_t eventPeriodicCreate(UAVObjEvent* ev, UAVObjEventCallback cb, struct pios_queue *queue, uint16_t periodMs)
+{
+	PeriodicObjectList* objEntry;
+	// Get lock
+	PIOS_Recursive_Mutex_Lock(mutex, PIOS_MUTEX_TIMEOUT_MAX);
+	// Check that the object is not already connected
+	LL_FOREACH(objList, objEntry)
+	{
+		if (objEntry->evInfo.cb == cb &&
+			objEntry->evInfo.queue == queue &&
+			objEntry->evInfo.ev.obj == ev->obj &&
+			objEntry->evInfo.ev.instId == ev->instId &&
+			objEntry->evInfo.ev.event == ev->event)
+		{
+			// Already registered, do nothing
```

Comment:
If the entry already exists but has a different callback period, we should just update the period (or if it doesn't have a period, delete the old entry and create a new one etc.). Unless you want the same callback with multiple different periods (which doesn't make sense).